File: alamos/go/trace.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package alamos

import (
	"context"

	"github.com/synnaxlabs/x/config"
	"github.com/synnaxlabs/x/errors"
	"github.com/synnaxlabs/x/override"
	"github.com/synnaxlabs/x/validate"
	"go.opentelemetry.io/otel/propagation"
	oteltrace "go.opentelemetry.io/otel/trace"
)

// Span is a span in a trace.
type Span interface {
	// Key returns the key provided when the span was created.
	Key() string
	// Error records the given error as an error on the span, setting the span's status
	// to Error if the error is non-nil. If exclude is provided, the status will only be
	// set if the error is not one of the excluded errors.
	Error(err error, exclude ...error) error
	// Status sets the span's status.
	Status(status Status)
	// EndWith combines Error and End, and also returns the given err unmodified.
	EndWith(err error, exclude ...error) error
	// End ends the span.
	End()
}

// TracingConfig is the configuration for a Tracer.
type TracingConfig struct {
	// OtelProvider sets the open telemetry tracing provider used to create spans.
	// [REQUIRED]
	OtelProvider oteltrace.TracerProvider
	// OtelPropagator sets the open telemetry propagator used to propagate spans across
	// process boundaries.
	// [REQUIRED]
	OtelPropagator propagation.TextMapPropagator
	// Filter is a function that is called to determine if a span should be created for
	// the given key and environment. If the filter returns false, the span will not be created.
	Filter EnvironmentFilter
}

var (
	_ config.Config[TracingConfig] = (*TracingConfig)(nil)
	// DefaultTracingConfig is the default configuration for a Tracer.
	DefaultTracingConfig = TracingConfig{
		Filter: ThresholdEnvFilter(EnvironmentBench),
	}
)

// Validate implements config.Config.
func (c TracingConfig) Validate() error {
	v := validate.New("alamos.tracing_config")
	validate.NotNil(v, "otel_provider", c.OtelProvider)
	validate.NotNil(v, "otel_propagator", c.OtelPropagator)
	validate.NotNil(v, "filter", c.Filter)
	return v.Error()
}

// Override implements config.Config.
func (c TracingConfig) Override(other TracingConfig) TracingConfig {
	c.OtelProvider = override.Nil(c.OtelProvider, other.OtelProvider)
	c.OtelPropagator = override.Nil(c.OtelPropagator, other.OtelPropagator)
	c.Filter = override.Nil(c.Filter, other.Filter)
	return c
}

// Tracer provides tracing functionality, and is one of the core components of
// Instrumentation. Tracer's should not be used on their own, and instead should
// be used as part of Instrumentation. To creat a Tracer, use NewTracer and pass
// it in a call to alamos.New using the WithTracer option.
type Tracer struct {
	config      TracingConfig
	_otelTracer oteltrace.Tracer
	meta        InstrumentationMeta
}

// NewTracer initializes a new devTracer using the given configuration. If no configuration
// is provided, NewTracer will return a validation error. If you want a no-op devTracer,
// simply use a nil pointer.
func NewTracer(configs ...TracingConfig) (*Tracer, error) {
	cfg, err := config.New(DefaultTracingConfig, configs...)
	if err != nil {
		return nil, err
	}
	return &Tracer{config: cfg}, nil
}

// Debug starts a span at the debug level with the given key. If the context is already
// wrapped in a span, the span will be a child of the existing span.
func (t *Tracer) Debug(ctx context.Context, key string) (context.Context, Span) {
	return t.Trace(ctx, key, EnvironmentDebug)
}

// Prod starts a span at the production level. If the context is already wrapped in a
// span, the span will be a child of the existing span.
func (t *Tracer) Prod(ctx context.Context, key string) (context.Context, Span) {
	return t.Trace(ctx, key, EnvironmentProd)
}

// Bench starts a span at the benchmark level. If the context is already wrapped in a
// span, the span will be a child of the existing span.
func (t *Tracer) Bench(ctx context.Context, key string) (context.Context, Span) {
	return t.Trace(ctx, key, EnvironmentBench)
}

// Trace wraps the given context in a span with the given key and level. If the context
// is already wrapped in a span, the span will be a child of the existing span.
func (t *Tracer) Trace(ctx context.Context, key string, env Environment) (context.Context, Span) {
	if t == nil || !t.config.Filter(env, key) {
		return ctx, nopSpan{}
	}
	spanKey := t.meta.extendPath(key)
	ctx, otel := t.otelTracer().Start(ctx, spanKey)
	return ctx, span{
		key:  spanKey,
		otel: otel,
	}
}

func (t *Tracer) otelTracer() oteltrace.Tracer {
	if t._otelTracer == nil {
		t._otelTracer = t.config.OtelProvider.Tracer(t.meta.Key)
	}
	return t._otelTracer
}

// Transfer transfers the trace from the source context to the target context.
// Transfer should be used sparingly, and is typically only used when preventing
// the propagation of context cancellation but the preservation of tracing:
//
//	def myFunc(ctx context.Context) {
//		// Transfer the trace so the given context cant be used for cancellation,
//		ctx = alamos.Transfer(ctx, context.Background())
//		// Will never return.
//		<-ctx.Done()
//	}
func (t *Tracer) Transfer(source, target context.Context) context.Context {
	return oteltrace.ContextWithSpan(target, oteltrace.SpanFromContext(source))
}

func (t *Tracer) child(meta InstrumentationMeta) (nt *Tracer) {
	if t != nil {
		nt = &Tracer{meta: meta, config: t.config}
	}
	return
}

type span struct {
	otel oteltrace.Span
	key  string
}

var _ Span = span{}

// Key implements Span.
func (s span) Key() string { return s.key }

// Error implements Span.
func (s span) Error(err error, exclude ...error) error {
	if err == nil {
		return err
	}
	s.otel.RecordError(err)
	if !errors.IsAny(err, exclude...) {
		s.Status(StatusError)
	}
	return err
}

// Status implements Span.
func (s span) Status(status Status) {
	s.otel.SetStatus(status.otel(), "")
}

// End implements Span.
func (s span) End() {
	s.otel.End()
}

// EndWith implements Span.
func (s span) EndWith(err error, exclude ...error) error {
	_ = s.Error(err, exclude...)
	s.End()
	return err
}

// nopSpan is a span that does nothing.
type nopSpan struct{}

func (s nopSpan) Key() string { return "" }

// Error implements Span.
func (s nopSpan) Error(err error, _ ...error) error { return err }

// Status implements Span.
func (s nopSpan) Status(_ Status) {}

// End implements Span.
func (s nopSpan) End() {}

// EndWith implements Span.
func (s nopSpan) EndWith(err error, _ ...error) error { return err }

File: alamos/ts/src/trace.ts
// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import {
  type AttributeValue,
  context,
  propagation,
  type Span as OtelSpan,
  SpanStatusCode,
  type Tracer as OtelTracer,
} from "@opentelemetry/api";
import { type destructor } from "@synnaxlabs/x";

import {
  type Environment,
  type EnvironmentFilter,
  envThresholdFilter,
} from "@/environment";
import { Meta } from "@/meta";

/** Carrier is an entitty that can carry trace metadata across process bounds */
export type Carrier = Record<string, string>;

/** Function that executes under the given span */
export type SpanF = (span: Span) => unknown;

/**
 * Tracer wraps an opentelemetry tracer to provide an opinionated intreface
 * for tracing within the Synnax stack.
 */
export class Tracer {
  private meta: Meta = Meta.NOOP;
  private readonly otelTracer?: OtelTracer;
  private readonly filter: EnvironmentFilter;

  constructor(
    tracer?: OtelTracer,
    filter: EnvironmentFilter = envThresholdFilter("debug"),
  ) {
    this.otelTracer = tracer;
    this.filter = filter;
  }

  child(meta: Meta): Tracer {
    const t = new Tracer(this.otelTracer, this.filter);
    t.meta = meta;
    return t;
  }

  debug(key: string): destructor.Destructor;
  debug<F extends SpanF>(key: string, f: F): ReturnType<F>;

  /**
   * Starts a new span in the debug environment. If a span already exists in the
   * current context, it will be used as the parent span.
   *
   * @param key - The name of the span.
   * @param f - The function to run under the span.
   * @returns A span that tracks program execution. If the Tracer's environment
   * rejects the 'debug' environment or the Tracer is noop, a NoopSpan is returned.
   */
  debug<F extends SpanF>(key: string, f?: F): ReturnType<F> | destructor.Destructor {
    return this.trace(key, "debug", f);
  }

  bench(key: string): destructor.Destructor;
  bench<F extends SpanF>(key: string, f: F): ReturnType<F>;

  /**
   * Starts a new span in the bench environment. If a span already exists in the
   * current context, it will be used as the parent span.
   *
   * @param key - The name of the span.
   * @param f - The function to run under the span.
   * @returns A span that tracks program execution. If the Tracer's environment
   * rejects the 'bench' environment or the Tracer is noop, a NoopSpan is returned.
   */
  bench<F extends SpanF>(key: string, f?: F): ReturnType<F> | destructor.Destructor {
    return this.trace(key, "bench", f);
  }

  prod(key: string): destructor.Destructor;
  prod<F extends SpanF>(key: string, f: F): ReturnType<F>;

  /**
   * Starts a new span in the prod environment. If a span already exists in the
   * current context, it will be used as the parent span.
   *
   * @param key - The name of the span.
   * @param f - The function to run under the span.
   * @returns A span that tracks program execution. If the Tracer's environment
   * rejects the 'prod' environment or the Tracer is noop, a NoopSpan is returned.
   */
  prod<F extends SpanF>(key: string, f?: F): ReturnType<F> | destructor.Destructor {
    return this.trace(key, "prod", f);
  }

  trace(key: string, env: Environment): destructor.Destructor;
  trace<F extends SpanF>(key: string, env: Environment, f: F): ReturnType<F>;
  trace<F extends SpanF>(
    key: string,
    env: Environment,
    f?: F,
  ): ReturnType<F> | destructor.Destructor;

  /**
   * Stars a new span with the given key and environment. If a span already
   * exists in the current context, it will be used as the parent span.
   *
   * @param key - The name of the span.
   * @param env - The environment to run the span under.
   * @param f - The function to run under the span.
   * @returns A span that tracks program execution. If the Tracer's environment
   * rejects the provided span or the Tracer is noop, a NoopSpan is returned.
   */
  trace<F extends SpanF>(
    key: string,
    env: Environment,
    f?: F,
  ): ReturnType<F> | destructor.Destructor {
    const skip = this.meta.noop || !this.filter(env) || this.otelTracer == null;
    if (f == null) {
      if (skip) return () => {};
      const span = new _Span(key, this.otelTracer.startSpan(key));
      span.start();
      return () => span.end();
    }

    if (skip) return f(new NoopSpan(key)) as ReturnType<F>;
    return this.otelTracer.startActiveSpan(key, (otelSpan) => {
      const span = new _Span(key, otelSpan);
      const result = f(span);
      return result as ReturnType<F>;
    });
  }

  /**
   * Injects metadata about the current trace into the provided carrier. This metadata
   * can be parsed on the other side of a network or IPC request to allow the trace to
   * propagate across services.
   *
   * @param carrier - The carrier to inject the metadata into.
   */
  propagate(carrier: Carrier): void {
    if (this.meta.noop) return;
    const ctx = context.active();
    propagation.inject(ctx, carrier, {
      set: (carrier, key, value) => {
        carrier[key] = value;
      },
    });
  }

  /** Tracer implementation that does nothing */
  static readonly NOOP = new Tracer();
}

/** A span in a trace that can be used to track function execution */
export interface Span {
  /**
   * The key identifying the span. This is the name of the key
   * passed into the tracing method combined with the path of the
   * instrumentation that started the span. For example, take the
   * instrumentation titled 'synnax' and call to trace with 'test.
   * The span key would be 'synnax.test'.
   */
  key: string;
  /**
   * If the error is not null, records the error in the span and sets
   * its status to error.
   */
  recordError: (error?: Error | null) => void;
  /**
   * Sets the given key-value pair as an attribute on the span.
   */
  set: (key: string, value: AttributeValue) => void;
}

export class _Span implements Span {
  key: string;
  private readonly otel: OtelSpan;

  constructor(key: string, span: OtelSpan) {
    this.key = key;
    this.otel = span;
  }

  set(key: string, value: AttributeValue): void {
    this.otel.setAttribute(key, value);
  }

  recordError(error?: Error | null): void {
    if (error == null) return;
    this.otel.recordException(error);
    this.otel.setStatus({ code: SpanStatusCode.ERROR });
  }

  start(): void {
    performance.mark(`alamos.trace.start.${this.key}`);
  }

  end(): void {
    try {
      performance.mark(`alamos.trace.end.${this.key}`);
      const duration = performance.measure(
        `alamos.trace.duration.${this.key}`,
        `alamos.trace.start.${this.key}`,
        `alamos.trace.end.${this.key}`,
      );
      this.set("duration", duration.duration);
      this.otel.end();
    } catch (e) {
      console.error(e);
    }
  }
}

/** Span implementation that does nothing */
export class NoopSpan implements Span {
  key: string;

  constructor(key: string) {
    this.key = key;
  }

  set(): void {}

  recordError(_?: Error | null): void {}
}
